[ci] Speedup beam search unit tests with fixtures for LLM

stnie opened the pull request:
## Description
CI/refactor: Speedup beam search unit tests with fixtures for LLMs, prompts and expected outputs

- Introduced pytest fixtures for input prompts and expected outputs to improve test readability and maintainability.
- Refactored the test function to utilize these fixtures, simplifying the parameter list and enhancing clarity.
- Introduced pytest fixture for LLM to skip rebuilding LLM for each test
- disabled threadleak to prevent failure, due to LLM not being shutdown after each test
- Updated assertions to ensure correct behavior of the beam search output shapes.

- updated small issues related to beam search
- Throw NotImplemented Error, when beam search is executed with CUDA Graph
- Fix Indexing Error in cum_log_probs


## Test Coverage

test_beam_search.py 

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Added a runtime check to prevent using CUDA Graphs with beam search, improving error handling and stability.
  * Corrected cumulative log probability handling for multi-beam scenarios to ensure accurate results.

* **Tests**
  * Refactored beam search tests to use reusable fixtures, reducing redundancy and improving test clarity.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

stnie committed:
refactor: enhance beam search unit tests with fixtures for LLM, prompts and expected outputs

- Introduced pytest fixtures for input prompts and expected outputs to improve test readability and maintainability.
- Refactored the test function to utilize these fixtures, simplifying the parameter list and enhancing clarity.
- Introduced pytest fixture for LLM to skip rebuilding LLM for each test
- disabled threadleak to prevent failure, due to LLM not being shutdown after each test
- Updated assertions to ensure correct behavior of the beam search output shapes.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #11292](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11292/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #11292](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11292/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8350 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8350/) completed with status: 'SUCCESS'

stnie committed:
chore: prevent CUDA Graph + beam search combination from being used

- Added a NotImplementedError to raise an exception if CUDA Graph is enabled with a max beam width greater than 1, indicating that this combination is not yet supported.
- Corrected the indexing for cumulative log probabilities in the sampler to ensure proper data access.
- Updated the LLM fixture in unit tests to use a dynamic max batch size based on input prompts, improving memory management and preventing potential data access issues.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #11956](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11956/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #11956](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11956/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8873](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8873/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/8873/test-results/rerun_report.html) for details.

Funatiq commented on tests/unittest/_torch/test_beam_search.py:
```diff
@@ -7,87 +7,105 @@
+@pytest.fixture(scope="module")
+def fixed_params():
+    return {
+        "kvcache_config": KvCacheConfig(max_tokens=10000),
```
Why not hardcode this in the `llm` fixture?

stnie replied:
I moved the KvCacheConfig into the llm fixture

Funatiq commented on tests/unittest/_torch/test_beam_search.py:
```diff
@@ -7,87 +7,105 @@
+                assert len(beam.logprobs) == sampling_params.max_tokens
+            else:
+                assert len(beam.logprobs) == 0
+            if num_output_beams == fixed_params["max_beam_width"]:
```
Do we need this condition? Shouldn't the the check also pass if `num_output_beams < fixed_params["max_beam_width"]`?

omera-nv reviewed (approved):
LGTM test-wise: the tests are now 48secs instead of 10mins, great work :)

stnie committed:
chore: updated test_beam_search.py
- enabled similarity check if less than max_beam_width outputs are returned.
- moved KvCacheConfig into the llm fixture

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #12198](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12198/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #12198](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12198/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #9060](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/9060/) completed with status: 'FAILURE'

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #12295](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12295/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #12295](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12295/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #9131](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/9131/) completed with status: 'SUCCESS'
